### PR TITLE
feature: deterministic random_split

### PR DESCRIPTION
**Pull request by artemisart**
## 🚀 Feature
Option to provide a seed (random_state) for random_split() like the sklearn API https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html.

## Motivation
Useful for deterministic sampling & reproducible data generation (easily, without affecting the PRNG for other uses).
See #32467

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
I don't think this is the right way to do it. Instead, I'd suggest passing in an optional Generator and then using that to generate the `randperm`. That gives users more flexibility in how they want `random_split` to be deterministic.

Also, the change needs a test.

**Comment by ezyang** [comment]
@pbelevich, since you're are working on RNGs, I assigned you to also take a look at this PR.

**Comment by artemisart** [comment]
@ezyang 
> I don't think this is the right way to do it. Instead, I'd suggest passing in an optional Generator and then using that to generate the `randperm`. That gives users more flexibility in how they want `random_split` to be deterministic.

How can I generate the randperm from a generator? Should I set the global state from the generator? And the randperm does not accept any generator parameter so I don't see how to do it.

**Comment by ezyang** [comment]
pass the generator to the invocation of randperm. 

```
>>> torch.randperm(1, "asdf")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: randperm() received an invalid combination of arguments - got (int, str), but expected one of:
 * (int n, torch.Generator generator, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 * (int n, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
```

there might be an unrelated doc bug where we don't document that generators are accepted.

**Comment by artemisart** [comment]
Indeed it's not in the docs nor __doc__, but it's much better with a generator!  I still kept random_state for 'compatibility' with train_test_split (and easier usage) but I can remove it if you prefer.  I will add some tests soon.

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -260,16 +260,25 @@ def __len__(self):
...
-def random_split(dataset, lengths):
+def random_split(dataset, lengths, random_state=None, generator=None):
     r"""
     Randomly split a dataset into non-overlapping new datasets of given lengths.
+    Optionally fix the random seed for reproducible results.
 
     Arguments:
         dataset (Dataset): Dataset to be split
         lengths (sequence): lengths of splits to be produced
+        random_state (int): seed used by the random number generator
```
Comment (ezyang) [inline torch/utils/data/dataset.py]: doc the generator? :)

File: torch/utils/data/dataset.py [hunk 2]
```diff
@@ -260,16 +260,25 @@ def __len__(self):
         return len(self.indices)
 
 
-def random_split(dataset, lengths):
+def random_split(dataset, lengths, random_state=None, generator=None):
```
Comment (ezyang) [inline torch/utils/data/dataset.py]: Any time you wrote `random_split(d, l, state)`, I could write instead `random_split(d, l, torch.Generator.manual_seed(state))`. The Pythonic way is that there is one obvious way to do it. So I suggest getting rid of the `random_state` argument entirely.

**Review by ezyang — COMMENTED** [review]
Looking forward to the tests.

File: torch/utils/data/dataset.py (hunk 2 above)
Comment (artemisart) [reply torch/utils/data/dataset.py]: Oh I hadn't realised `manual_seed` returns the generator so it's possible to keep a one-liner.

**Pull request closed without merging**